### PR TITLE
ENH: named agg with multiple columns

### LINKED ISSUE #29268: Named aggregations with multiple columns
Since `pandas 0.25.0` we have [named aggregations](https://pandas.pydata.org/pandas-docs/stable/whatsnew/v0.25.0.html#groupby-aggregation-with-relabeling). 

Which works fine if you do aggregations on single columns. But what if you want to apply aggregations over multiple columns:

**example**:

```
# example dataframe

df = pd.DataFrame(np.random.rand(4,4), columns=list('abcd'))
df['group'] = [0, 0, 1, 1]

          a         b         c         d  group
0  0.751462  0.572576  0.192957  0.921723      0
1  0.070777  0.801548  0.601678  0.344633      0
2  0.112964  0.361984  0.416241  0.785764      1
3  0.380045  0.486494  0.000594  0.608759      1

# aggregations on single columns
f.groupby('group').agg(
             a_sum=('a', 'sum'),
             a_mean=('a', 'mean'),
             b_mean=('b', 'mean'),
             c_sum=('c', 'sum'),
             d_range=('d', lambda x: x.max() - x.min())
)

          a_sum    a_mean    b_mean     c_sum   d_range
group                                                  
0      0.947337  0.473668  0.871939  0.838150  0.320543
1      0.604149  0.302074  0.656902  0.542985  0.057681
```
But what if we want to calculate the `a.max() - `b.max()` while aggregating. That does not seem to work. For example, something like this would make sense:
```
df.groupby('group').agg(
    diff_a_b=(['a', 'b'], lambda x: x['a'].max() - x['b'].max())
)
```
---
So is it possible to do named aggregations on multiple columns? If not, is this in the pipeline for future releases?

### PR DESCRIPTION
**Pull request by erfannariman**
- [ ] closes #29268
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by jreback** [comment]
pls add tests always

in fact tests should be done before any code changes to make sure they behave properly

**Comment by erfannariman** [comment]
Thanks for the heads up @jreback , actually I was hoping to get some feedback which was the reason I pushed this as a draft while developping to get some comments. Some parts are challenging since I am not sure if I make the right decisions.

**Review by ivanovmg — COMMENTED** [review]
Can you add the test based on the example in the issue https://github.com/pandas-dev/pandas/issues/29268 where you demonstrated aggregation ``a.max() - b.max()``?
This will be a good starting point for testing.
Then we can think of covering edge cases.

File: pandas/core/aggregation.py [hunk 1]
```diff
@@ -727,7 +727,10 @@ def agg_dict_like(
             elif (
                 isinstance(selected_obj, ABCDataFrame) and k not in selected_obj.columns
             ):
-                raise KeyError(f"Column '{k}' does not exist!")
+                if is_list_like(k):
+                    for c in k:
+                        if c not in selected_obj.columns:
+                            raise KeyError(f"Column '{c}' does not exist!")
```
[the remark this answers is not in the document]
Comment (erfannariman) [reply pandas/core/aggregation.py:733]: Yes, good point, added. Thanks.

[the remark this answers is not in the document]
Comment (erfannariman) [reply pandas/core/aggregation.py:733]: What would you think of :

```python
not_found = [c for c in k if c not in selected_obj.columns]
raise KeyError(f"Column(s) {', '.join(not_found)} do(es) not exist!")
```

**Comment by erfannariman** [comment]
@jreback @ivanovmg I made some progress, but I am at the point that I need to pass the DataFrame as a obj to the aggregator instead of a Series so we can do the aggregation over the columns.

**Comment by jreback** [comment]
@erfannariman status here?

**Comment by erfannariman** [comment]
> @erfannariman status here?

This went stale, but is definitely one I would like very much to finish. I just think it gets quite complex on some points where I could use some guidance. Will take asap to clean up, merge master etc so it's updated.

**Comment by lithomas1** [comment]
@erfannariman Can you merge master/update? Thanks.

**Comment by simonjayhawkins** [comment]
Thanks @erfannariman for the PR. closing as stale. ping if you want to continue.

**Comment by erfannariman** [comment]
> Thanks @erfannariman for the PR. closing as stale. ping if you want to continue.

Started to work on this PR again, can you re-open? Thanks @simonjayhawkins

**Comment by jreback** [comment]
this is quite old, happen to reopen if actively worked on.

**Pull request closed without merging**